Add KEP on a new taint-manager

yuanchen8911 opened the pull request:
<!-- 
	Please use the following format when naming your PR
	< Issue Number >:< Issue Description >
	e.g. KEP-000: adding beta graduation criteria
	
	Avoid using phrases like `fixes #NNNN` in the description
	unless the pull request is to change the KEP status to 
	implemented or KEP has been deprecated.
-->

<!-- short description of work done in PR e.g. updating milestone, adding new KEP, adding test requirements… -->  
- One-line PR description:  Decouple `taint-manager` form `NodeLifecycleController` to support custom taint-based pod eviction

<!-- link to the k/enhancements issue -->
- Issue link: TODO

<!-- other comments or additional information -->
- Other comments:  https://github.com/kubernetes/kubernetes/issues/115779

yuanchen8911 commented:
sig/node
sig/scheduling

yuanchen8911 commented:
/cc @atiratree @Huang-Wei

yuanchen8911 commented:
/sig node scheduling

kerthcet commented:
FYI: may this https://github.com/kubernetes/kubernetes/pull/115840 related? I didn't read  the KEP.

atosatto commented:
> FYI: may this https://github.com/kubernetes/kubernetes/pull/115840 related? I didn't read the KEP.

@kerthcet we see this as a preliminary step aiming to remove deprecated features in `NodeLifecycleController` to enable the work proposed in this KEP

yuanchen8911 commented:
@soltysh @kow3ns, @alculquicondor,  thanks for your the discussions on the proposal in today's `sig-apps` meeting. Your review on the KEP will be very appreciated.  @Huang-Wei will provide feedback from the `sig-scheduling` perspective. Thanks!

alculquicondor commented:
ownership discussion https://github.com/kubernetes/kubernetes/issues/116771

yuanchen8911 commented:
Hi @alculquicondor, @ravisantoshgudimetla mentioned you wanted us to re-organize the KEP by moving some items to the alternative implementation during the sig-apps meeting. Can you please let me know the changes you are suggesting? I'd be happy to address them. Thanks!

yuanchen8911 commented on keps/sig-apps/independent-taint-manager/README.md:
```diff
@@ -0,0 +1,310 @@
+
+Toleration cannot provide a flexible and general enough extension mechanism to handle different Pod eviction and deletion cases to meet various workload requirements. 
+
+* The toleration mechanism is not flexible enough to handle complex stateful applications and workloads that require specific coordination of pod eviction, e.g. checkpointing for certain classes of machine-learning workloads to guarantee data integrity before deletion.
```
> Can you explore as an alternative ways of making it flexible enough to accomodate your usecase?

Thanks. A summary and discussion of alternatives were added to the KEP.

yuanchen8911 commented:
Revised the KEP and moved it to `sig-scheduling`, who owns it now.   /cc @soltysh @Huang-Wei

yuanchen8911 commented:
/cc @atosatto

atosatto commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,305 @@
+    * Custom mutating webhook may interfere with the admission plugin that applies the default tolerations if they are trying to apply tolerations for similar taints.
+    * Toleration update in Webhooks can also cause conflicts with older version of pods especially during upgrades and we need to ensure a toleration update is valid.
+
+* **Enhance the existing toleration APIs**
+    * Introduce a new Toleration effect `EvictIfAllowed` similar to `PreferNoSchedule` with additional fields on each Toleration to specify a reference to another CR/API object for how to reach the external system that governs if eviction is allowed similar to webhooks. Similar to the communication between API Server and admission webooks, `NoExecTaintManager` and an external service can establis ...
+    * Add RBAC for tolerations API on who can add what tolerations. Otherwise, a user can add tolerations to their workloads(pods) and make every pod land on the node to cause a noisy neighbor and even DoS problem in the worst case.
```
I think this is a separated problem than the one we are aiming to solve with this proposal. While there is definitely room for extending the Kubernetes API to support a more graceful form of `NoExecute` respecting the PodDisruptionBudget API (we had a session on this at the Contributor Summit in Amsterdam last week), the goal of this proposal is to ensure operators can extend Kubernetes by providing custom logic handling Taint Based Eviction to change the default behaviour vs having to patch Kubernetes Controller Manager.

yuanchen8911 replied:
> I think this is a separated problem than the one we are aiming to solve with this proposal. While there is definitely room for extending the Kubernetes API to support a more graceful form of `NoExecute` respecting the PodDisruptionBudget API (we had a session on this at the Contributor Summit in Amsterdam last week), the goal of this proposal is to ensure operators can extend Kubernetes by providing custom logic handling Taint Based Eviction to change the default behaviour vs having to patch Kubernetes Controller Manager.

Added the comment to the discussion of alternatives.

yuanchen8911 replied:
The use and enhancement of tolerations by injecting tolerations for NoExecute is a not viable at scale because of the factors described above.

Although extending the Kubernetes API to support a more graceful form of NoExecute respecting the PodDisruptionBudget API is possible, this proposal aims to enable operators to extend Kubernetes by providing custom logic for handling taint-based evictions. This approach changes the default behavior rather than requiring patching the Kubernetes Controller Manager.

yuanchen8911 replied:
Splitting the components is not only for cleanup, but also a prerequisite for enhancing the pod eviction manager, whether it's the native or a custom controller.

Regarding your point about enhancing the default controller, we believe it would be helpful, but we also recognize that it may not be practical or even possible to have a general controller that meets all needs in different use cases, as described in the motivation and use cases.

We will make further clarification in the proposal. 

Thank you for taking the time to review our KEP and share your thoughts.

SergeyKanzhelev replied:
this is what was discussed at SIG Node meeting as well. It will be great to capture improvements planned or desired and pain points and explain how decoupling will help to address those.

yuanchen8911 replied:
We summarize the benefits and improvements by decoupling `TaintManager` from `NodeLifecycleController`.

- A more logical code structure with two separate controllers for tainting nodes and evicting pods respectively. 
- Greater flexibility to support customized taint-based eviction to meet the specific needs of different workloads than the use of the toleration-based mechanism, which is less flexible and harder to customize.
- Ability to enable/disable the default `TaintManager` by setting or unsetting the "--controllers" flag of `kube-controller-manager`.

ravisantoshgudimetla replied:
I think what @yuanchen8911  is suggesting is decoupling will make it easy for the future changes which are listed above. For example, once the split happens, it would be easier to make a HTTP call to external system which has more context and knowledge of the workloads from the taint controller(from an implementation standpoint).

yuanchen8911 replied:
> > the goal of this proposal is to ensure operators can extend Kubernetes by providing custom logic handling Taint Based Eviction to change the default behaviour.
> 
> You can already do that by disabling the controller.
> 
> I don't see any problem with splitting the existing controller into two. This is a rather welcomed cleanup.
> 
> I disagree that the motivation for splitting it is so that people can write their own. We should be looking for opportunities to enhance the existing controller and API so that people don't feel the need to fork it.

I agree splitting will enable to enhance and improve the existing controllers, but it will also provide opportunities for cluster admins to write their own taint managers like scheduler plugins.

yuanchen8911 commented:
/assign @Huang-Wei

Huang-Wei commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,309 @@
+
+A new `NoExecuteTaintManager` is created a first-class controller managed by` kube-controller-manager`. It’s implementation is based on the current taint-manager  `controller/nodelifecycle/taint-manager.go`.  A new flag is added 
+
+`// NoExecuteTaintManagerConfiguration contains elements describing NoExecuteTaintManagerConfiguration`
+`type NoExecuteTaintManagerConfiguration struct {`
+`// enables the default NoExecuteTaintManager. MUST be synced with the`
+`// corresponding flag of the kube-apiserver.`
+`EnableNoExecuteTaintManager bool`
+`  ...`
+`}`
```
Could you reformat this section?

Huang-Wei commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,309 @@
+
+A new `NodeLifeCycleManager` is implemented by removing taint-manager related code from `controller/nodelifecycle/node_lifecycle_controller.go`.
+
+A new `NoExecuteTaintManager` is created a first-class controller managed by` kube-controller-manager`. It’s implementation is based on the current taint-manager  `controller/nodelifecycle/taint-manager.go`.  A new flag is added 
```
We shouldn't say "a new flag", do we? as it will be a top-level controller controlled by `--controllers`.

```suggestion
A new `NoExecuteTaintManager` is created as a first-class controller managed by `kube-controller-manager`. Its implementation is based on the current taint-manager `controller/nodelifecycle/taint-manager.go`.
```

Huang-Wei commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
````diff
@@ -0,0 +1,309 @@
+        return nil, false, nil
+    }`
+
+```
````
```suggestion
```go
```

Huang-Wei commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,309 @@
+
+The creation and startup of the default taint-manager is performed by `kube-controller-manager` . 
+
+`func startNoExecuteTaintManager(ctx context.Context, controllerContext ControllerContext) (controller.Interface, bool, error) {`
+`    if !controllerContext.ComponentConfig.NoExecuteTaintManager.EnableNoExecuteTaintManager {
+        return nil, false, nil
+    }`
```
Please format this snippet.

Huang-Wei commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,309 @@
+
+### Test Plan
+
+* Unit and integration tests
+    * Add nodes taints
+    * Act on `NoExecute` to delete running pods
+    * Respect taint tolerations 
+* Verify passing the existing E2E and conformance tests.
+    * Taint a node 
+    * Evict the correct set of running pods
```
You may need to detail the tests section. There is a updated version requiring documenting the current code coverage, etc. You can refer to other latest KEP for references.

yuanchen8911 replied:
How about 

* Unit tests
    * Enable `TaintManager`
    * Disable `TaintManager`
    * Add `NoExecute` taints
    * Act on `NoExecute` to delete running pods
    * Respect taint tolerations
* Integration tests
    * Enable `TaintManager`, add `NoExecute` taints to evict the correct set of running pods on nodes
    * Disable `TaintManager`, add `NoExecute` has no impact on running pods
* E2E tests
    * Verify passing the existing E2E and conformance tests.

Huang-Wei commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,309 @@
+
+#### Alpha
+
+* Support `kube-controller-manaager` flag `controllers=-taint-manager` to disable the default taint-manager.
```
Is the `-` semantics really what we're going to implement? I'd assume disabling a default controller can be achieved by providing a full list of controller names.

If this is a goal of this KEP, I highly recommend to document it in the goals section.

yuanchen8911 replied:
Yes, it's already listed in the goal section.

ravisantoshgudimetla replied:
I guess this is side effect of making taint-manager a separate controller. This is not the goal?

yuanchen8911 replied:
By decoupling the taint-manager from nodelifecyclecontroller, we can just use existing controller-manager flag to disable the default taint-manager instead of using a separate flag. I agree it's not the main goal. I'll remove it.

Huang-Wei commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,309 @@
+* Starting kubernetes 1.27, with the removal of the flag `enable-taint-manager`, replacing the default `NoExecute` taint-manager with a custom one would be more challenging. `NoExecute` taints can disable the no-execute taint manager, but such disabling can only be achieved through a mutating webhook on all pods, and it can interact negatively with the upstream `DefaultTolerations` webhook.
+* The requirement of custom taint-based eviction is not unique. It exists already in Kubernetes, where we call it `FullyDistributionMode`. When all nodes are unhealthy, `NodeLifeCycleController` chooses not to apply `NoExecute` taints properly. In a real-world case, the integrator may choose to define the disruption rate/budget, etc.
+* `NodeCycleController` combines two independent functions: adding a pre-defined set of `NoExecute` taints to unhealthy nodes and performing pod eviction on arbitrary `NoExecute` taints. While  `NodeLifecycle` controller responsibility is primarily on Node unhealthy-zone, `TaintManger` is concerned with actioning the taints added by `NodeLifecycle` controller by performing Pods eviction. De-coupl ...
+
```
It'd be good add a Notes section here to explain that the term `TaintManger` and `NoExecuteTaintManager` will be used interchangeably in this doc.

yuanchen8911 replied:
Added a note section.

`taint-manager`,  `TaintManger` and `NoExecuteTaintManager`  are used interchangeably in this doc.

Huang-Wei commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,309 @@
+
+## Design Details
+
+### Current implementation
```
The term `current` should be avoided to use: https://kubernetes.io/docs/contribute/style/style-guide/#avoid-statements-that-will-soon-be-out-of-date.

yuanchen8911 replied:
Deleted the word `Current` and changed the description to `In version 1.27 and prior of Kubernetes, ...`

Huang-Wei commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,309 @@
+
+<p align="center"><image src="new-controllers.png" width=60%></p>
+
+### Custom taint-manager
```
I don't think we should mention this part in the design details, as it's out-of-tree option.

yuanchen8911 replied:
The section and diagram were removed from the KEP.

Huang-Wei commented on keps/sig-scheduling/3902-decoupled-taint-manager/kep.yaml:
```diff
@@ -0,0 +1,47 @@
+milestone:
+  alpha: "v1.28"
+  beta: "v1.29"
+  stable: "v1.30"
```
there needs to be at least 2 versions from beta to stable.

yuanchen8911 replied:
Removed `stable`.

yuanchen8911 commented on keps/prod-readiness/sig-scheduling/3902.yaml:
```diff
@@ -0,0 +1,3 @@
+kep-number: 3902
```
The issue number is 3902. https://github.com/kubernetes/enhancements/issues/3902

yuanchen8911 commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,309 @@
+# KEP-3902:  Decouple Taint-based Pod Eviction from Node LifeCycle Controller 
```
The issue number is https://github.com/kubernetes/enhancements/issues/3902.

yuanchen8911 commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,309 @@
+
+#### Story 2
+
+A cluster orchestrator can design and implement a centralized pod eviction manager that handles all pod eviction and deletions on different `NoExecute`  taints (`Unreachable`, `NotReady` and others) and any custom taints, e.g., acting on the conditions defined in `descheduler`, such as node overloading, etc.
```
done.

yuanchen8911 commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,309 @@
+
+## Summary
+
+In Kubernetes, `NodeLifeCycleController` applies predefined `NoExecute` taints (e.g., `Unreachable`, `NotReady`) when the nodes are determined to be unhealthy. After the nodes get tainted, the `taint-manager` does its due diligence to start deleting running pods on those nodes based on `NoExecute` taints, which can be added by anyone.
```
`taint-manager` --> `TaintManager`

alculquicondor commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,308 @@
+* Whether or not evict a stateful pod with local storage on `NoExecute` can depend on the actual taint conditions, workload status and other conditions. The decision may require information from the external systems and controllers, which have more knowledge and context about the workloads.  A custom `taint-manager`  will be able to control how and when a pod is evicted or deleted, including maki ...
+* Starting kubernetes 1.27, with the removal of the flag `enable-taint-manager`, replacing the default `NoExecute` taint-manager with a custom one would be more challenging. `NoExecute` taints can disable the no-execute taint manager, but such disabling can only be achieved through a mutating webhook on all pods, and it can interact negatively with the upstream `DefaultTolerations` webhook.
+* The requirement of custom taint-based eviction is not unique. It exists already in Kubernetes, where we call it `FullyDistributionMode`. When all nodes are unhealthy, `NodeLifecycleController` chooses not to apply `NoExecute` taints properly. In a real-world case, the integrator may choose to define the disruption rate/budget, etc.
+* `NodeLifecycleController` combines two independent functions: adding a pre-defined set of `NoExecute` taints to unhealthy nodes and performing pod eviction on arbitrary `NoExecute` taints. While  `NodeLifecycleController`'s responsibility is primarily on Node unhealthy-zone, `TaintManager` is concerned with actioning the taints added by `NodeLifecycleController` by performing Pods eviction. De- ...
```
This should be the top motivation and, from my point-of-view, the only motivation.

yuanchen8911 replied:
OK, moved the point to the top and restructured the motivations as
1. Splitting the controllers will improve maintainability and enable more flexible extension and customization of taint-based eviction.
2. There are increasing needs to extend and customize the default taint-manager in practice.

alculquicondor commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,308 @@
+A new `NoExecuteTaintManager` is created a first-class controller managed by` kube-controller-manager`. It’s implementation is based on the current taint-manager  `controller/nodelifecycle/taint-manager.go`.  A new flag is added 
+
+`// NoExecuteTaintManagerConfiguration contains elements describing NoExecuteTaintManagerConfiguration`
+`type NoExecuteTaintManagerConfiguration struct {`
```
why do we need this, in addition to the `--controllers` flag?

yuanchen8911 replied:
You are right. We don't need any new flag. Removed it.

alculquicondor commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,308 @@
+### Goals
+
+* Move taint-based eviction implementation out of `NodeLifecycleController` into a separate and independent taint-manager.
+* Using the `--controllers`  flag of `kube-controller-manager` to enable and disable the default `taint-manager` to support customizable taint-manager controller. 
```
IIUC, this is about any controller

ravisantoshgudimetla replied:
Only https://github.com/kubernetes/enhancements/pull/3901/files#diff-489db758e0c012ad5037ad23ca8bdc76e11ec4f7117f53f7d057f3878343fc67R89 is relevant here

alculquicondor commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,308 @@
+
+* **Leverage the existing tolerations for `NoExecute` taints**: operators opt-out from taint-based eviction by injecting tolerations for NoExecute taints. It has a few major disadvantages over an independent `TaintManager`.
+    * It will require to either inject tolerations to all pods via mutating webhooks and/or changing the manifests of all the running workloads. 
+    * It will require either injecting tolerations to every single pod via mutating webhooks and/or changing the manifests of all the running workloads.
```
same point as above?

alculquicondor commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,308 @@
+    * Toleration update in Webhooks can also cause conflicts with older version of pods especially during upgrades and we need to ensure a toleration update is valid.
+
+* **Enhance the existing toleration APIs**
+    * Introduce a new Toleration effect `EvictIfAllowed` similar to `PreferNoSchedule` with additional fields on each Toleration to specify a reference to another CR/API object for how to reach the external system that governs if eviction is allowed similar to webhooks. Similar to the communication between API Server and admission webooks, `NoExecTaintManager` and an external service can establis ...
```
There are two APIs that allow extensibility of kubernetes without disabling default behavior:
- finalizers https://kubernetes.io/docs/concepts/overview/working-with-objects/finalizers/
- scheduling gates https://kubernetes.io/docs/concepts/scheduling-eviction/pod-scheduling-readiness/

Could you consider similar approaches?

I don't think the proposal in this paragraph is detailed enough, hence why it might not feel sufficient for your use case.

ravisantoshgudimetla replied:
> Could you consider similar approaches?


The only closest thing that I can think of is scheduler extender(similar to entity which is outside of Kube system).

>  hence why it might not feel sufficient for your use case.

Which use-case are you referring to here? Splitting taint-manager?

alculquicondor replied:
Splitting the controllers is not a use-case. It's an implementation detail.

I'm not sure what use case the authors are trying to solve. I suppose it's close to story 1.

> The only closest thing that I can think of is scheduler extender

That goes along the way of "I want to fully override the component", which is the same as this proposal.

alculquicondor replied:
Right, that could be a valid avenue towards providing flexibility. However, not all users have access to the control plane to configure the webhook. So a more generic API or something more similar to finalizers could be more usable.

ravisantoshgudimetla replied:
Agreed. I think the changes that @yuanchen8911 is suggesting are at the API level and he is also against having a webhook. However the implementation details may vary. Having said that, I guess we're debating about how we want to implement a gap in Kube in the alternate section of this KEP. The gap and how it needs to be addressed warrants another KEP. So, if you've any objections with the wording or the content(including the alternatives section), please let us know and @yuanchen8911 can make changes to move forward.

alculquicondor replied:
I already expressed my point: I'm in favor of the refactoring. I disagree with the Motivation and Stories. Given the current motivations, I consider that the Alternatives section hand-waves that improvements to the API wouldn't work without going into much detail.

ravisantoshgudimetla replied:
I agree with your assessment on the motivation and use-cases section. 

However in the alternatives section, https://github.com/kubernetes/enhancements/pull/3901/files#diff-489db758e0c012ad5037ad23ca8bdc76e11ec4f7117f53f7d057f3878343fc67R301 the KEP states that the goal is not to change the API at this point of time. 

I believe https://github.com/kubernetes/enhancements/pull/3901/files#diff-489db758e0c012ad5037ad23ca8bdc76e11ec4f7117f53f7d057f3878343fc67R299 is relevant only to using webhooks not extending API. We feel that extending API is a good alternative but it is a larger change and it needs addressing few gaps that we mentioned. We will change the wording if that is what concerns you. Let me know if you have any other concerns.

yuanchen8911 replied:
> I also disagree with that stance.

What would be your recommendation, Aldo?  Any specific things you want us to add to the KEP?

ravisantoshgudimetla replied:
I guess, the point of contention here is `the side-effect of splitting the controllers is that it allows cluster operators to build custom TaintManagers and disable the default TaintManager`. @yuanchen8911 went ahead and explained why it is beneficial to split from that standpoint. As I mentioned earlier, we cannot seriously explore alternatives in this KEP as we may digress from the original point. So, better to cut short and explain this KEP on separation of concerns and tell the side-effect of this split is people can disable default `TaintManager`.

alculquicondor replied:
I like that approach better. It doesn't give false expectations to other people following the KEP.

yuanchen8911 replied:
Updated the Non-Goals to clarify the code reorganizations the main goal of the KEP as below.

### Non-Goals

The main focus of the KEP is on the separation of concerns between `NodeLifecycleController` and `TaintManager`, which will improve code organization and enable future extensions to the `TaintManager`. While it is true that the separation allows cluster operators to build custom `TaintManager` and disable the default `TaintManager`, this is a side-effect of the change rather than the main goal. the KEP emphasizes the benefits of the separation of concerns and briefly mention the option to disable the default TaintManager as a potential side-effect.

* Extend and enhance the current built-in APIs for `TaintManager` customization, including additional fields, more flexible communication protocols, webhooks and CRD references, etc.
* Actual extension and enhancement  of the current `TaintManager`.

alculquicondor reviewed (commented):
I don't think the current KEP is doing a good job of exploring alternatives to disabling existing functionality.
I left a couple of APIs for inspiration.

ravisantoshgudimetla commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,308 @@
+
+### Non-Goals
+
+* Any change to the behavior of the current taint-manager
```
Yup. Remove this from non-goals. This KEP is just for the split which improves the code maintenance.

yuanchen8911 replied:
Agree "`This KEP is just for the split which improves the code maintenance and makes future extensions easier`." .  The non-goal here means extension and customization of the default taint-manager behavior is out of the scope of the proposal.

yuanchen8911 commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,308 @@
+
+## Design Details
+
+In version 1.27 and prior of Kubernetes, `NoExecuteTaintManager` is part of `NodeLifecycleController`. It’s created and run if `enable-taint-manager` flag is true.  The flag is deprecated and will be removed in kubernetes 1.27. 
```
What we meant here is there will be no way to disable the default pod eviction behavior with the flag removed.

alculquicondor replied:
It always has been, before or after the flag was removed.

yuanchen8911 replied:
Before the flag is removed, we can at least disable the default TaintManager.

alculquicondor replied:
But then pods are immediately deleted based on unhealthy Node conditions, unless you disable the entire NodeLifecycleController.

See this code in 1.26 https://github.com/kubernetes/kubernetes/blob/4dfe380379bc9b3c39f8b041a95d12b4e6ac0cf3/pkg/controller/nodelifecycle/node_lifecycle_controller.go#L1329-L1335

My point is that we didn't lose the ability to "disable pod deletion" because we removed `enable-taint-manager`. That was not the purpose of the flag.

yuanchen8911 commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,308 @@
+## Motivation
+
+* Whether or not evict a stateful pod with local storage on `NoExecute` can depend on the actual taint conditions, workload status and other conditions. The decision may require information from the external systems and controllers, which have more knowledge and context about the workloads.  A custom `taint-manager`  will be able to control how and when a pod is evicted or deleted, including maki ...
+* Starting kubernetes 1.27, with the removal of the flag `enable-taint-manager`, replacing the default `NoExecute` taint-manager with a custom one would be more challenging. `NoExecute` taints can disable the no-execute taint manager, but such disabling can only be achieved through a mutating webhook on all pods, and it can interact negatively with the upstream `DefaultTolerations` webhook.
```
Yes, but without the flag, it will be extremely difficult to change the default pod eviction behavior. By splitting the flag, we can reuse the controller-manager's flag `--controllers` by adding or removing taintmanager.

yuanchen8911 commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,309 @@
+
+#### Story 4
+
+A pod may be stuck in terminating state for various reason, e.g.,  `kubelet` is not up and `AppServer` cannot reach `kubelet`, etc. The hanging deletion may prevent new pods starting because the persistent volume claim is not released. A custom taint-manager can perform a forceful deletion as needed.
```
Typo, `APIServer`. Fixed it.

yuanchen8911 commented:
@alculquicondor @ravisantoshgudimetla , thanks a lot for the detailed reviews and very helpful comments. I made changes to  the `Motivations` and `Goals` sections. I hope they have addressed some of your concerns. 

IIRIC, we all agree splitting the controllers is beneficial as it can improve code maintainability and provide opportunities to enhance the existing `TaintManager`.  We are debating

- Is writing custom `TaintManager` a main use case and motivation?  While the proposed refactoring will certainly make the future enhancement of the default `TaintManager` easier, we'd argue it will also provide opportunities to write custom `TaintManager` like scheduler plugins.  The use stories demonstrate why a custom `TaintManager` is needed.  One size does not fit all.  The default `TaintManager` will not be able to handle all cases. Even if we can have a generic `TaintManager` and provide customization via extended APIs,  the proposed splitting is the first step and prerequisite for creating such a manager.

-  Do we have better alternatives and is API extension&enhancement an option?  
   - It's reasonable to argue that API change is harder.
   - The proposed splitting will make API extension and enhancement easier.  
   - API enhancement is not the goal of the proposal. 
   - In the alternatives, we demonstrate that the existing toleration API is not enough for the use cases and that splitting the controllers is necessary to address these limitations. 
   - It's still worth considering other alternatives in the future.

We'll revise the use cases and alternatives sections. 

What do you you think about it? @alculquicondor?

/cc @Huang-Wei  @atosatto @ddebroy

alculquicondor commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,300 @@
+
+### Goals
+
+* Move taint-based eviction implementation out of `NodeLifecycleController` into a separate and independent taint-manager to support flexible extension and customization of node taint-based eviction.
```
```suggestion
* Move taint-based eviction implementation out of `NodeLifecycleController` into a separate and independent taint-manager to enhance separation of concerns and maintainability.
```

alculquicondor replied:
The current wording can give the false impression that anyone can implement an extension, which is not true. Only users who have the ability to replace or configure control plane components can.

yuanchen8911 replied:
Good point, we'll clarify that a only the cluster administrators can do it.

Wel112 replied:
It will monitor these information changes. If a node has not updated its status for a long time, then we can assume that an exception has occurred in this node, and some fault tolerance processing is needed to safely remove the pod on this node. These Pods are rebuilt on other nodes.

alculquicondor commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,300 @@
+* `NodeLifecycleController` combines two independent functions: adding a pre-defined set of `NoExecute` taints to unhealthy nodes and performing pod eviction on arbitrary `NoExecute` taints. While  `NodeLifecycleController`'s responsibility is primarily on Node unhealthy-zone, `TaintManager` is concerned with actioning the taints added by `NodeLifecycleController` by performing Pods eviction. De- ...
+* There are increasing needs to customize and extend the default node taint-based pod evictions in practice.
+  * Whether or not evict a stateful pod with local storage on `NoExecute` can depend on the actual taint conditions, workload status and other conditions. The decision may require information from the external systems and controllers, which have more knowledge and context about the workloads.  A custom `taint-manager`  will be able to control how and when a pod is evicted or deleted, including ma ...
+  * Starting kubernetes 1.27, with the removal of the flag `enable-taint-manager`, replacing the default `NoExecute` taint-manager with a custom one would be more challenging. `NoExecute` taints can disable the no-execute taint manager, but such disabling can only be achieved through a mutating webhook on all pods, and it can interact negatively with the upstream `DefaultTolerations` webhook.
```
This cannot be a motivation, as `enable-taint-manager=false` would still evict pods, just through a different (even less flexible) mechanism

yuanchen8911 replied:
Yes, `descheduler` or other controllers can still evict pods, but `enable-taint-manager=false` will disable `NoExecute`-based pod eviction performed by the default `TaintManager`. We will remove it if you are strongly against it. 
 /cc @atosatto

alculquicondor replied:
I'm not talking about descheduler. I'm talking about kube-controller-manager itself. When you set `enable-taint-manager=false`, kcm can evict pods just based on Node conditions.

atosatto replied:
> When you set enable-taint-manager=false, kcm can evict pods just based on Node conditions.

@alculquicondor this is correct. However the functionality was deprecated and has been removed in https://github.com/kubernetes/kubernetes/pull/115840 as part of the effort to improve code-health in NodeLifeCycleController which this KEP is part of (as per your comment in https://github.com/kubernetes/enhancements/pull/3901#discussion_r11937486320).

The main benefit from de-coupling taint-based eviction from NodeLifeCycleController is that it would enable end users who have the possibility to control their control-plane configuration, to disable entirely taint-based eviction in their clusters without having to mutate all of the Pods via tolerations. 

While I do agree that there is room for extending the current built-in APIs to add (e.g. web-hooks) this should be part of a  different proposal and not in-scope of the current KEP which is purely focusing on re-organising the code.

alculquicondor replied:
That is correct, it's removed.

But the KEP here is trying to say that because we removed the flag, there is less control over eviction. That is not true, thus can't be part of the motivation.

alculquicondor replied:
> the main benefit from de-coupling taint-based eviction from NodeLifeCycleController is that it would enable end-users who have the possibility to control their control-plane configuration, to disable entirely taint-based eviction in their clusters without having to mutate all of the Pods via tolerations.

I disagree. Most end-users won't be able to disable the taint-based-eviction. And if they want to have some behavior similar to NoExecute taints with extra functionality, they still have to re-implement what kcm offers. So this is not a good extension mechanism.

> While I do agree that there is room for extending the current built-in APIs to add (e.g. web-hooks) this should be part of a different proposal and not in-scope of the current KEP which is purely focusing on re-organising the code.

I fully agree. That's what I have listed as one of the options to move forward with this KEP: to focus on re-organising.

yuanchen8911 replied:
Updated the alternative with a focus on the code refactoring too. 

## Alternatives

Keeping the code as-is is an option, but it will make future extension of `TaintManager` harder and less flexible

yuanchen8911 commented:
Hi @alculquicondor, based on your feedback and suggestions, we have simplified the motivation and user stories. The main objective now is to split the two controllers in order to achieve a more flexible extension and enhancement of the TaintManager. Could you please take a look? Thank you so much! /cc @ravisantoshgudimetla  @Huang-Wei  @atosatto  @ddebroy

alculquicondor commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,271 @@
+1. `NodeLifecycleController` monitors node health and adds `NotReady` and `Unreachable` taints to nodes 
+2. `NoExecuteTaintManager` watches for node and pod updates,  and performs `NoExecute` taint based pod eviction
+
+The existing  kube-controller-manager flag `--controllers` will be used to optionally disable the `NoExecuteTaintManager`.
```
```suggestion
The existing  kube-controller-manager flag `--controllers` can be used to optionally disable either controller.
```

alculquicondor commented on keps/sig-scheduling/3902-decoupled-taint-manager/kep.yaml:
```diff
@@ -0,0 +1,46 @@
+# The following PRR answers are required at alpha release
+# List the feature gate name and the components for which it must be enabled
+feature-gates:
+  - name: controllers 
```
this should be the name of the feature gate that you plan to use

yuanchen8911 replied:
How about
```
feature-gates:
  - name: SeparateTaintManager
    components:
      - kube-controller-manager
      - kube-apiserver
disable-supported: true
```

alculquicondor replied:
why would be apiserver be affected?

alculquicondor commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,271 @@
+
+<p align="center"><image src="new-controllers.png" width=60%></p>
+
+### Implementation
```
Will there be a feature gate guarding the split? In other words, can operators use the combined controller, if they find any issues with the split, without having to rollback?

alculquicondor commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
````diff
@@ -0,0 +1,271 @@
+  }  
+```
+
+### Test Plan
````
Please follow the template https://github.com/kubernetes/enhancements/tree/master/keps/NNNN-kep-template#test-plan

yuanchen8911 replied:
Updated the test plan following the new template.

alculquicondor commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,271 @@
+
+#### Story
+
+While this split is more of a kubernetes developer focused change, an unintended positive effect of this change is that it allows cluster-admins to develop custom `TaintManager`s and disable the default `TaintManager`. 
```
```suggestion
While this split is more of a kubernetes developer focused change, an effect of this change is that it allows cluster-admins to develop custom `TaintManager`s and disable the default `TaintManager`. 
```

yuanchen8911 commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,271 @@
+
+While this split is more of a kubernetes developer focused change, an unintended positive effect of this change is that it allows cluster-admins to develop custom `TaintManager`s and disable the default `TaintManager`. 
+
+The default `TaintManager`, because of the implementation rigidity, may not be suitable for every cluster operator. We except users to come up with custom `TaintManager`s and this KEP may exacerbate the problem of `bring your own taint manager`. However, discussing the use-cases for extending `TaintManager` or custom `TaintManger`s is beyond the scope of this KEP. 
```
Typo of `expect`.

yuanchen8911 commented:
Hi @alculquicondor, I updated the PR based on the recent discussions and suggestions.
1. Further clarified the main goal of the KEP is code reorganization for more flexible extension of `TaintManager` and how to extend and customize the default `TaintManager` is out of the scope.  
2. Updated `Test Plan` following the new template.
3. Fixed various issues and typos.

Thanks! 

 /cc @Huang-Wei @ravisantoshgudimetla @atosatto @ddebroy

alculquicondor commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
````diff
@@ -0,0 +1,292 @@
+
+The creation and startup of the default taint-manager is performed by `kube-controller-manager` . 
+
+```
````
I would remove the code snippet. It's not relevant in the KEP.

yuanchen8911 replied:
OK, deleted it.

alculquicondor commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,292 @@
+Kubernetes already has a good coverage of node `No-Execute` eviction. We will add tests only for the changed code.
+
+#####  Unit tests
+- `pkg/controller/taintmanager`: 2023-05-23 - 0%
+- `pkg/controller/apis/config/v1alpha1`: 2023-05-23 - 0%
```
what is the coverage for the existing code?

yuanchen8911 replied:
-  Enable and disable the new TaintManager
- The new TaintManager acts on node taints properly

alculquicondor commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,292 @@
+
+### Graduation Criteria
+
+#### Alpha
```
Since there are no new API fields, we could skip alpha and go to beta directly.
However, PRR reviewer might have a different opinion.

yuanchen8911 replied:
Updated it.

alculquicondor commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,292 @@
+#### Alpha
+
+* Support `kube-controller-manager` flag `--controllers=-taint-manager` to disable the default taint-manager.
+* Unit and integration tests completed and passed.
```
E2E?

yuanchen8911 replied:
Fixed it.

alculquicondor commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,292 @@
+
+*This section must be completed when targeting alpha to a release.*
+
+* **How can this feature be enabled / disabled in a live cluster?**
```
Mention the feature gate

wojtek-t replied:
Please update to the new KEP template (as mentioned above) - the feature-gate is then required explicitly :)

alculquicondor commented:
mostly LGTM.
You should start requesting approval from SIG Node and PRR.

alculquicondor commented:
LGTM from sig scheduling

yuanchen8911 commented:
> LGTM from sig scheduling

Hi @SergeyKanzhelev, @wojtek-t, the KEP has  gone though extensive reviews and approved by sig-scheduling. Would you review it for sig-node and PRR please? Thanks!

/cc @alculquicondor @Huang-Wei

soltysh commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,266 @@
+  - Enable and disable the new `TaintManager`
+  - The new `TaintManager` acts on node taints properly
+- `pkg/controller/apis/config/v1alpha1`: 2023-05-23 - 0% 
+  - Use the split `TaintManager` or the old combined controller
```
here you should list the existing tests, so I'd expect here `pkg/controller/nodelifecycle` and its subdirs.

soltysh commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,266 @@
+  - Use the split `TaintManager` or the old combined controller
+
+#####  Integration tests
+ No integration tests are required. This feature is better tested with e2e tests.  
```
Why? Integration tests allow you to have a more strict environment that you have a full control on.

yuanchen8911 replied:
Sounds good.  Added integration tests.

soltysh commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,266 @@
+
+### Upgrade / Downgrade Strategy
+
+We’ll need to upgrade `kube-controller-manager` and` node-lifecycle-controller` and add a new controller `taint-manager`. Accordingly, a downgrade will revert to the previous version of `kube-controller-manager` and `node-lifecycle-controller` and remove the new added `taint-manager` . There should be no impact on the other components during upgrades and downgrades since user-exposed behavior sho ...
```
This doesn't make sense. kube-controller-manager is one process, which embeds several controllers (node-lifecycle-controller and taint-manager, when added). The upgrade and downgrade description here should point out that this only affects kcm, so there are no expected problems during upgrade and downgrade. 
Unless you start disabling the built-in taint-manager and using a custom built one, in which case you're approach will be different, but that case is again non-goal, so I'm not sure it's worth describing it here.

Huang-Wei replied:
I think we can reword in the perspective of user's angle:

For a user that doesn't use this feature, no matter which kind of upgrade/downgrade path is performed, the eventual result should be that both "applying NoExecute taints on nodes upon conditions" and "evict pods upon NoExecute taints" function well. (they don't care about the underlying code restructured or not)

For a user that uses this feature, downgrade would mean (in-tree) taint-manager will be enabled, no matter it's previously enabled or not; upgrade would mean (in-tree) taint-manager will be enabled, but can be disabled. We don't need to mention out-of-tree taint-manager, as 1) that's no different from an external controller, and 2) nothing can stop a customer runs duplicated or faulty taint-manager.

yuanchen8911 replied:
How about the following

A feature gate `SeparateTaintManager`  enables and disables the new feature. When the feature is turned on, a user can use `kube-controller-manager`'s flag to disable the default `TaintManager`.

alculquicondor replied:
I think the point is that what you are saying is part of the Design, or the question "how can the user disable this feature".
But it's not part of the upgrade/downgrade strategy.
The answer is "N/A" for this KEP

soltysh commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,266 @@
+* **Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?**
+    * Yes, we will need to rack back `kube-controller-manager` and `node-lifecycle-controller`. 
+* **What happens if we re-enable the feature if it was previously rolled back?**
+    * Upgrade `kube-controller-manager` and `node-lifecycle-controller` and add the new `taint-manager`.
```
This has a similar problem like my other comment.

yuanchen8911 replied:
Changed it to  "Use the feature gate `SeparateTaintManager` to enable or disable the feature."

soltysh commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,266 @@
+* **What happens if we re-enable the feature if it was previously rolled back?**
+    * Upgrade `kube-controller-manager` and `node-lifecycle-controller` and add the new `taint-manager`.
+* **Are there any tests for feature enablement/disablement?**
+    * Planned for Alpha release
```
Alpha or beta?

yuanchen8911 replied:
Alpha

soltysh commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,266 @@
+    * Disable the default taint-manager
+    * Create custom taint-based pod eviction
+* **How can someone using this feature know that it is working for their instance?**
+    * Node taints and taint-based pod eviction work as usual.
```
This won't allow you to differentiate between old and new behavior. If all goes as planned admins and users should not see the difference, right?

soltysh commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,266 @@
+* **How can someone using this feature know that it is working for their instance?**
+    * Node taints and taint-based pod eviction work as usual.
+* **What are the reasonable SLOs (Service Level Objectives) for the enhancement?**
+    * Taint-based pod eviction latency
```
Metric?

yuanchen8911 replied:
Taint-based pod eviction latency (e.g., p99) should be less than x seconds.

soltysh commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,266 @@
+
+In version 1.27 and prior of Kubernetes, `NoExecuteTaintManager` is part of `NodeLifecycleController` and created and run if `enable-taint-manager` flag is true. 
+
+<p align="center"><image src="current-nodelifecycle-controller.png" width=50%></p>
```
`![Current Controllers](current-nodelifecycle-controllers.png)`

soltysh reviewed (commented):
Left some comments, although it doesn't fall under sig-apps umbrella that much, but most importantly if you're planning to push this through for 1.28 ping sig-scheduling leads to opt-in https://github.com/kubernetes/enhancements/issues/3902 into the release

yuanchen8911 commented on keps/sig-scheduling/3902-decoupled-taint-manager/kep.yaml:
```diff
@@ -0,0 +1,47 @@
+  - name: SeparateTaintManager 
+    components:
+      - kube-controller-manager
+      - kube-apiserver
```
deleted `kube-apiserver`

yuanchen8911 commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,266 @@
+
+The main focus of the KEP is on the separation of concerns between `NodeLifecycleController` and `TaintManager`, which will improve code organization and enable future extensions to `TaintManager`. While it is true that the separation allows cluster operators to build custom `TaintManager` and disable the default `TaintManager`, this is a side-effect of the change rather than the main goal. the K ...
+
+* Extend and enhance the current built-in APIs for customization of node taint-based eviction, including additional fields, more flexible communication protocols, webhooks and CRD references, etc.
+* Actual extension and enhancement  of the current `TaintManager`.
```
Would it be better to have them as part of the Goals instead if we add "Do Not" ?

- Do not introduce extra maintenance burden or regression 
- Do not introduce incompatible behavior to NodeLifecycleController or TaintManager

Huang-Wei replied:
Nope, the bullets are under `Non-Goals`, so use a positive tone is correct.

yuanchen8911 replied:
Are you suggesting we put the two bullets under `Non-Goals`? It sounds weird to me. If they are Non-Goals, would it meany we don't care if the changes will introduce extra maintenance burden and incompatible behavior? Also, these requirements should be self-evident (needless to say) for most changes if not all. 

```
Introduce extra maintenance burden or regression after the code reorganization
Introduce incompatible behavior of NodeLifecycleController or TaintManager
```

Huang-Wei replied:
> Are you suggesting we put the two bullets under Non-Goals?

Yes.

> Also, these requirements should be self-evident (needless to say) for most changes if not all.

It can be debating. But IMO, code restructuring  in this KEP is non-trivial work, and hence may cause regression/incompatibility issue, so listing them as Non-Goals doesn't sound redundant.

yuanchen8911 commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,266 @@
+- [ ]  (R) Production readiness review approved
+- [ ]  "Implementation History" section is up-to-date for milestone
+- [ ]  User-facing documentation has been created in [kubernetes/website](https://git.k8s.io/website), for publication to [kubernetes.io](https://kubernetes.io/)
+- [ ]  Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
```
@Huang-Wei , would you suggest which boxes I should check? Thanks.

Huang-Wei replied:
Just check the boxes that are already completed. Like "Enhancement issue in release milestone", "Impl. history", etc.

yuanchen8911 commented:
Hi @soltysh , @Huang-Wei , thanks a lot for the detailed reviews and helpful suggestions. I made another revision to address the comments. Can you take another look? Thanks again!

soltysh commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,271 @@
+* **Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?**
+    * Yes, we will need to rack back `kube-controller-manager` and `node-lifecycle-controller`. 
+* **What happens if we re-enable the feature if it was previously rolled back?**
+    * Use the feature gate `SeparateTaintManager` to enable or disable the feature.
```
This still doesn't answer the question, what can happen if you re-enable the feature after a previous rollback? For example: you upgraded the cluster and started using this feature and something went wrong, so you rollback to previous version to get the problem resolved. Afterwards you're again turning it on? What can go wrong? 

In your case I think it's safe to say that there should be no unexpected problems.

yuanchen8911 replied:
Yes, there should be no problem. Updated the proposal.
```
The new feature will work as usual. 
```

soltysh commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,271 @@
+* **What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?**
+    * Taint-based pod eviction latency
+* **Are there any missing metrics that would be useful to have to improve observability of this feature?**
+    * Taint-based pod eviction latency 
```
Can you point out the exact metric name?

yuanchen8911 replied:
Will change it to `No` as this change should have minimal performance impact on the pod eviction behavior.

soltysh commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,271 @@
+* **How can someone using this feature know that it is working for their instance?**
+    * Node taints and taint-based pod eviction work as usual. Admins and users should not see any different behavior if everything works as planned.
+* **What are the reasonable SLOs (Service Level Objectives) for the enhancement?**
+    * Taint-based pod eviction latency (e.g., P99) should be less than x seconds.
```
What is "x seconds"?

yuanchen8911 replied:
How about `N/A`? This changes should have minimal performance impact on the pod eviction behavior.

soltysh commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,271 @@
+
+* **How can an operator determine if the feature is in use by workloads?**
+    * Disable the default taint-manager
+    * Create custom taint-based pod eviction
```
Given the change is deep in the core, there's no easy way to know that, is it?

yuanchen8911 replied:
If the feature gate `SeparateTaintManage`r used?

Huang-Wei commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,270 @@
+Kubernetes already has a good coverage of node `No-Execute` eviction. We will add tests only for the changed code.
+
+#####  Unit tests
+- `pkg/controller/taintmanager`: 2023-05-23 - 0%  
```
There is no pkg `pkg/controller/taintmanager`. You meant `pkg/controller/nodelifecycle/scheduler`?

yuanchen8911 replied:
I'm referring it to the proposed new `TaintManager`.

Huang-Wei commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,270 @@
+  - The new `TaintManager` acts on node taints properly
+- `pkg/controller/apis/config/v1alpha1`: 2023-05-23 - 0% 
+  - Test new configuration  
+- `pkg/controller/nodelifecycle`, `pkg/controller/nodelifecycle/scheduler`: 2023-05-23 - 100% 
```
100% doesn't look right:

<img width="910" alt="image" src="https://github.com/kubernetes/enhancements/assets/1425903/0d32aadb-edba-48ef-a6a8-93d32b453404">

I'd suggest putting separate pkg/controller/nodelifecycle and pkg/controller/nodelifecycle/scheduler

yuanchen8911 replied:
Made the suggested change.

Huang-Wei commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,270 @@
+* **Does enabling the feature change any default behavior?**
+    * No, the default node-life-cycle controller and taint-manager behavior will stay the same.
+* **Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?**
+    * Yes, we will need to rack back `kube-controller-manager` and `node-lifecycle-controller`. 
```
```suggestion
    * Yes. Once rolled back, `taint-manager` is enabled along with `node-lifecycle-controller`. 
```

Huang-Wei commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,270 @@
+* **What happens if we re-enable the feature if it was previously rolled back?**
+    * The feature will work as usual.
+* **Are there any tests for feature enablement/disablement?**
+    * Planned for Beta release
```
```suggestion
    * Appropriate tests have been added in the integration tests. See [Integration tests](#integration-tests) for more details.
```

yuanchen8911 commented:
@Huang-Wei , can you suggest how we should proceed to make the 1.28 release? Thanks!

https://github.com/kubernetes/enhancements/issues/3902#issuecomment-1575225809

Huang-Wei commented:
/approve on scheduling side.

Leaving another approval for @soltysh from sig-apps' perspective.

And after that, we will pursue PRR review from @wojtek-t .

soltysh commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,273 @@
+* **What are the reasonable SLOs (Service Level Objectives) for the enhancement?**
+    * N/A as the change will have minimal impact on the performance and there's no service associated with it. 
+* **What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?**
+    * N/A
```
I'd at least expect the metrics for both controllers' queues to maintain current levels, and that should be explicitly pointed out in this proposal.

yuanchen8911 replied:
How about "The performance of node taint-based eviction should remain the same level as before."?

soltysh commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,273 @@
+* **How can someone using this feature know that it is working for their instance?**
+    * Node taints and taint-based pod eviction work as usual. Admins and users should not see any different behavior if everything works as planned.
+* **What are the reasonable SLOs (Service Level Objectives) for the enhancement?**
+    * N/A as the change will have minimal impact on the performance and there's no service associated with it. 
```
Theoretically - yes it should not have impact, but how a cluster administrator can learn that? Think of both controllers like black-boxes and what would you look at from the outside that will give you assurance that the black-box still works as before the upgrade? See my other comment about queues' metrics

soltysh reviewed (commented):
From sig-apps pov this looks good, but the metrics in PRR section will need to be explicitly called out, since I'm pretty sure PRR review will ask just for that.

alculquicondor commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,273 @@
+
+### User Stories (Optional)
+
+#### Story
```
If the focus of the KEP is maintainability, we don't really need a story. This is confusing.

yuanchen8911 replied:
> If the focus of the KEP is maintainability, we don't really need a story. The paragraphs below are adding confusion.

How about moving it out of `Story` as the a topic level note? 

"While this split is for improving the code maintainability, an effect of this change is that it allows cluster-admins to extend and enhance the default `TaintManager` and even replace the default `TaintManager` with a custom implementation. However, discussing such use-cases for extending `TaintManager` or writing  custom `TaintManager` is beyond the scope of this KEP."

alculquicondor replied:
that sounds like a better location.

alculquicondor reviewed (commented):
You still need approval from SIG node, the other of NodeLifecycleController.

yuanchen8911 commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,273 @@
+
+## Design Details
+
+In Kubernetes version 1.27 and earlier, the `NoExecuteTaintManager` component is included within the `NodeLifecycleController`. It is created and executed only if the `enable-taint-manager` flag is set to true. 
```
OK, removed  the sentence "t is created and executed only if the `enable-taint-manager` flag is set to true. "

yuanchen8911 commented:
> You still need approval from SIG node, the other of NodeLifecycleController.

Hi @SergeyKanzhelev, this KEP has been reviewed extensively and approved by sig-scheduling (owner) and sig-apps. Can you please review and approve it for sig-node?  

The target is k8s 1.28 release and we are approaching the  enhancements freeze on 01:00 UTC Friday, 16th June 2023.

yuanchen8911 commented:
> /approve on scheduling side.
> 
> Leaving another approval for @soltysh from sig-apps' perspective.
> 
> And after that, we will pursue PRR review from @wojtek-t .

Hi @wojtek-t , the KEP has been approved by sig-scheduling and sig-apps. Can we get a PRR review please?  The target is k8s 1.28 release and the enhancements freeze is on 01:00 UTC Friday, 16th June 2023. Thank you so much!

SergeyKanzhelev commented on keps/sig-scheduling/3902-decoupled-taint-manager/kep.yaml:
```diff
@@ -0,0 +1,46 @@
+
+# The milestone at which this feature was, or is targeted to be, at each stage.
+milestone:
+  alpha: "v1.28"
```
please match the issue metadata. Is it going to go directly to beta?

yuanchen8911 replied:
Thanks! Corrected it.

SergeyKanzhelev commented on keps/sig-scheduling/3902-decoupled-taint-manager/kep.yaml:
```diff
@@ -0,0 +1,46 @@
+reviewers:
+  - "@Huang-Wei" 
+  - "@alculquicondor"
+approvers:
```
you can put me for sig node reviewer of approver

yuanchen8911 replied:
Added your name to reviewers.

SergeyKanzhelev commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,272 @@
+
+### Notes/Constraints/Caveats (Optional)
+
+### Risks and Mitigations
```
I guess the risk here is that it will be harder to implement scenarios that need both features to be in the same controller. Just dreaming out loud those might be:

1. Fast notification between these two controllers to minimize time between taint being applied and eviction started. 
2. Cancellation of Pods eviction.
3. Pod eviction logic un-tainting the Node based on some signal received during the evicton.

I think mitigation for all of them would be the fact that we likely don't want this behavior built-in. Worth mentioning I think

yuanchen8911 replied:
Good point! How about adding the following

"
Compared with a single combined controller, the risks of using two separate controllers include the following.
* Slightly increase the communication overhead from applying node taints to performing pod eviction.
* Make cancellation of pod eviction and un-tainting nodes harder.

It's a tradeoff between using the built-in simple controller and customizing the behavior to meet advanced needs.
"

SergeyKanzhelev replied:
> It's a tradeoff between using the built-in simple controller and customizing the behavior to meet advanced needs

I do not agree with the statement that built-in is necessarily simple and separation is the way to meet the advanced needs. Combined controller can be configureable. I think justification is more along the lines of cleaner design is better as we do not believe those scenarios are very important

yuanchen8911 replied:
> > It's a tradeoff between using the built-in simple controller and customizing the behavior to meet advanced needs
> 
> I do not agree with the statement that built-in is necessarily simple and separation is the way to meet the advanced needs. Combined controller can be configureable. I think justification is more along the lines of cleaner design is better as we do not believe those scenarios are very important

How about the following? 

`
While the performance impact of the split controllers is expected to be minimal, and pod eviction cancellation and un-tainting nodes are not common use cases, splitting  `NodeLifecycleController` promotes a cleaner design and enhances the overall maintainability and extensibility of in the long term.
`

yuanchen8911 replied:
Thanks, Sergey. May I assume the KEP has been approved by you (sig-node)?

SergeyKanzhelev commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,272 @@
+
+* Fix all reported bugs if any.
+
+### Upgrade / Downgrade Strategy
```
if somebody had the controller replaced, after the split the second one will stay? SO they will need to disable one more?

yuanchen8911 replied:
For downgrade, disabling the new feature using the feature gate will automatically disable any custom controller and use the current and default combined controller.

yuanchen8911 commented:
@SergeyKanzhelev , thanks for the comments! Can you take another look?

wojtek-t commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,278 @@
+
+### Version Skew Strategy
+
+## Production Readiness Review Questionnaire
```
Please update to the new KEP template (release team will ask for it anyway):
https://github.com/kubernetes/enhancements/blame/master/keps/NNNN-kep-template/README.md#L430

yuanchen8911 replied:
done

wojtek-t commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,278 @@
+    * The feature will work as usual.
+* **Are there any tests for feature enablement/disablement?**
+    * Planned for Beta release
+    * Appropriate tests have been added in the integration tests. See [Integration tests](#integration-tests) for more details.
```
These tests haven't been added, rather they are planned, right?

That being said, if I'm reading correctly, those tests are rather checking if things work fine when  FG is enabled. They are not trying to test the enablement/disablement - being "run k8s with FG enabled, restart controller-manager with FG disabled and check if things still work as expected (and vice-versa)"

Which BTW is probably fine here, given this is in-memory feature (so no state is really stored here).
So I would be ok with saying sth like:
"No enablement/disablement tests are needed since this is in-memory feature and just regular tests with the feature being enabled/disabled do the job."

yuanchen8911 replied:
updated as suggested.

wojtek-t commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,278 @@
+*This section must be completed when targeting beta graduation to a release.*
+
+* **How can a rollout or rollback fail? Can it impact already running workloads?**
+  A feature gate `SeparateTaintManager` controls whether to use the split `TaintManager` or the old combined `NodeLifecycleController`.
```
Please answer the question. Can it impact already running workloads?

yuanchen8911 replied:
Does this make sense?

"This is an opt-in feature, and it does not change any default behavior. Unless there is a bug in the implementation, a  rollout can not fail. If a rollout does fail, running workloads will not be evicted propertly on tainted nodes. We don't excpet a rollback can fail."

wojtek-t commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,278 @@
+  A feature gate `SeparateTaintManager` controls whether to use the split `TaintManager` or the old combined `NodeLifecycleController`.
+
+* **What specific metrics should inform a rollback?**
+  N/A
```
This isn't N/A for sure.

Some huge number of pod evictions? [Please add appropriate metric]

yuanchen8911 replied:
How about "A significantly changing number of pod evictions and/or a substantial increase in pod eviction latency"?

wojtek-t commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,278 @@
+  N/A
+
+* **Were upgrade and rollback tested? Was the upgrade→downgrade→upgrade path tested?**
+  They will be covered in unit and e2e tests.
```
Will they? We don't really have downgrade e2e tests running anywhere.

Please run them manually and describe the test case
[https://github.com/kubernetes/enhancements/pull/3658 is a nice example]

yuanchen8911 replied:
Added "manually verify the rollback and upgrade-downgrade-upgrade path will pass the e2e testing"

yuanchen8911 replied:
> The overall proposal makes sense to me (including going directly to Beta), but I have a bunch of comments about the PRR itself - PTAL

Thanks for the review, Wojciech! I've updated the doc to (hopefully) address the comments :-). Can you please take another look when you get a chance? Thanks again!

wojtek-t commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,278 @@
+* **How can an operator determine if the feature is in use by workloads?**
+    * It can be determined by if the feature gate `SeparateTaintManager` is used.
+* **How can someone using this feature know that it is working for their instance?**
+    * Node taints and taint-based pod eviction work as usual. Admins and users should not see any different behavior if everything works as planned.
```
How can I determine it as an operator running gazillions clusters?

Please specify a metric (number of pod evictions is probably a best one).

wojtek-t commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,278 @@
+* **What are the reasonable SLOs (Service Level Objectives) for the enhancement?**
+    * The performance of node taint-based eviction should remain the same level as before.
+* **What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?**
+    * The metrics for both `NodeLifecycleController` and `TaintManager`'s queues should stay the same levels as before.
```
What metrics? Please be specific.

yuanchen8911 replied:
How about "the number of pod evictions and pod eviction latency"?

wojtek-t commented on keps/sig-scheduling/3902-decoupled-taint-manager/README.md:
```diff
@@ -0,0 +1,278 @@
+* **What are other known failure modes?**
+    * No
+* **What steps should be taken if SLOs are not being met to determine the problem?**
+    * If the default or custom taint-manager is working or not. 
```
I don't understand it - can you please clarify?

yuanchen8911 replied:
Updated to 

If the pod eviction latency increases significantly, validate if the communication between `NodeLifecycleController` and `TaintManager` works. If the number of pod evictions is abnormal, run tests to verify the TaintManager works properly.

wojtek-t reviewed (commented):
The overall proposal makes sense to me (including going directly to Beta), but I have a bunch of comments about the PRR itself - PTAL

SergeyKanzhelev reviewed (approved):
/lgtm

lgtm from SIG Node.
